src/useWorker.ts

```diff
@@ -0,0 +1,104 @@
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable no-console */
+import React from 'react'
+import createWorkerBlobUrl from './lib/createWorkerBlobUrl'
+import WORKER_STATUS from './lib/status'
+
+type Options = {
+  timeout?: number;
+  dependencies: string[];
+}
+
+const PROMISE_RESOLVE = 'resolve'
+const PROMISE_REJECT = 'reject'
+const DEFAULT_OPTIONS: Options = {
+  timeout: undefined,
+  dependencies: [],
+}
+
+/**
+ *
+ * @param {Function} fn the function to run with web worker
+ * @param {Object} options useWorker option params
+ */
+export const useWorker = <T extends (...fnArgs: any[]) => any>(fn: T, options: Options = DEFAULT_OPTIONS) => {
+  const [workerStatus, setWorkerStatus] = React.useState<WORKER_STATUS>(WORKER_STATUS.PENDING)
+  const worker = React.useRef<Worker & { _url?: string }>()
+  const promise = React.useRef<{ [PROMISE_REJECT]?: (result: ReturnType<T> | ErrorEvent) => void;[PROMISE_RESOLVE]?: (result: ReturnType<T>) => void }>({})
+  const timeoutId = React.useRef<number>()
+
+  const killWorker = (status = WORKER_STATUS.PENDING) => {
+    if (worker.current && worker.current._url) {
+      worker.current.terminate()
+      URL.revokeObjectURL(worker.current._url)
+      promise.current = {}
+      worker.current = undefined
+      clearTimeout(timeoutId.current)
+      setWorkerStatus(status)
+    }
+  }
+
+  React.useEffect(() => () => {
+    killWorker()
+  }, [])
+
+  const generateWorker = () => {
+    const {
+      dependencies = DEFAULT_OPTIONS.dependencies,
+      timeout = DEFAULT_OPTIONS.timeout,
+    } = options
+    const blobUrl = createWorkerBlobUrl(fn, dependencies)
+    const newWorker: Worker & { _url?: string } = new Worker(blobUrl)
+    newWorker._url = blobUrl
+
+    newWorker.onmessage = (e: MessageEvent) => {
+      const [status, result] = e.data as [WORKER_STATUS, ReturnType<T>]
+
+      switch (status) {
+        case WORKER_STATUS.SUCCESS:
+          promise.current[PROMISE_RESOLVE]!(result)
+          killWorker(status)
+          break
+        default:
+          promise.current[PROMISE_REJECT]!(result)
+          killWorker(WORKER_STATUS.ERROR)
+          break
+      }
+    }
+
+    newWorker.onerror = (e: ErrorEvent) => {
+      promise.current[PROMISE_REJECT]!(e)
+      killWorker(WORKER_STATUS.ERROR)
+    }
+
+    if (timeout) {
+      timeoutId.current = setTimeout(() => {
+        killWorker(WORKER_STATUS.TIMEOUT_EXPIRED)
+      }, timeout) as any as number
```

Comment:
`as any as number` I never seen 2 `as` used like this before, out of curiosity what does this do?

Reply:
I just made a suggestion above, that should remove the need of this cast

Reply:
the reason for doing that is that "@types/node" is one of the dependencies, and "setTimeout" returns not a number in the Node, I've found an appropriate solution for this type stuff

Reply:
Resolved adding `window.setTimeout`

More info: https://stackoverflow.com/a/51040768/1477950